core/contextmenu_registry.js

```diff
@@ -0,0 +1,138 @@
+ * @typedef {{
+ *    block: (Blockly.BlockSvg|undefined),
+ *    workspace: (Blockly.WorkspaceSvg|undefined),
+ * }}
+ */
+Blockly.ContextMenuRegistry.Scope;
+
+/**
+ * A menu item as entered in the registry.
+ * @typedef {{
+ *    callback: function(*),
+ *    scopeType: !Blockly.ContextMenuRegistry.ScopeType,
+ *    displayText: ((function(!Blockly.ContextMenuRegistry.Scope):string)|string),
+ *    preconditionFn: function(!Blockly.ContextMenuRegistry.Scope):string,
+ *    weight: number,
+ *    id: string,
+ * }}
+*/
+Blockly.ContextMenuRegistry.RegistryItem;
+
+/**
+ * A menu item as presented to contextmenu.js.
+ * @typedef {{
+ *    text: string,
+ *    enabled: boolean,
+ *    callback: function(*),
+ *    scope: !Blockly.ContextMenuRegistry.Scope,
+ * }}
+ */
+Blockly.ContextMenuRegistry.ContextMenuOption;
+
+/**
+ * Registry of all registered RegistryItems, keyed by id.
+ * @type {!Object<string, Blockly.ContextMenuRegistry.RegistryItem>}
+ */
+Blockly.ContextMenuRegistry.registry_ = {};
+
+/**
+ * Registers a RegistryItem.
+ * @param {Blockly.ContextMenuRegistry.RegistryItem} item Context menu item to register.
+ * @throws {Error} if an item with the given id already exists.
+ */
+Blockly.ContextMenuRegistry.register = function(item) {
+  if (Blockly.ContextMenuRegistry.registry_[item.id]) {
+    throw Error('Menu item with id "' + item.id + '" is already registered.');
+  }
+  Blockly.ContextMenuRegistry.registry_[item.id] = item;
+};
+
+/**
+ * Unregisters a RegistryItem with the given id.
+ * @param {string} id The id of the RegistryItem to remove.
+ * @throws {Error} if an item with the given id does not exist.
+ */
+Blockly.ContextMenuRegistry.unregister = function(id) {
+  if (Blockly.ContextMenuRegistry.registry_[id]) {
+    delete Blockly.ContextMenuRegistry.registry_[id];
+  } else {
+    throw new Error('Menu item with id "' + id + '" not found.');
+  }
+};
+
+/**
+ * @param {string} id The id of the RegistryItem to get.
+ * @returns {Blockly.ContextMenuRegistry.RegistryItem} RegistryItem
+ * @throws {Error} if an item with the given id does not exist.
+ */
+Blockly.ContextMenuRegistry.getItem = function(id) {
+  if (Blockly.ContextMenuRegistry.registry_[id]) {
+    return Blockly.ContextMenuRegistry.registry_[id];
+  } else {
+    throw new Error('Menu item with id "' + id + '" not found.');
+  }
+};
+
+/**
+ * Gets the valid context menu options for the given scope type (e.g. block or workspace) and scope.
+ * Blocks are only shown if the preconditionFn shows they should not be hidden.
+ * @param {Blockly.ContextMenuRegistry.ScopeType} scopeType Type of scope where menu should be
+ *     shown (e.g. on a block or on a workspace)
+ * @param {Blockly.ContextMenuRegistry.Scope} scope Current scope of context menu
+ *     (i.e., the exact workspace or block being clicked on)
+ * @returns {!Array<!Blockly.ContextMenuRegistry.ContextMenuOption>} the list of ContextMenuOptions
```

Comment:
nit: `Array.<` (Blockly style)

Reply:
Done, and I think I caught the rest of them, but I have a question on how the team handles style guide recs that might change. The Closure compiler calls this "deprecated syntax" (https://github.com/google/closure-compiler/wiki/Types-in-the-Closure-Type-System under "Type Application") which is the only reference I could find to doing it this way. When something like that changes, do you prefer to keep using the syntax for uniformity, change it as you go, change it all at once?